src/bsp/platform/esp32/xi_bsp_io_net_esp32.c

```diff
@@ -0,0 +1,313 @@
+/* Copyright (c) 2003-2016, LogMeIn, Inc. All rights reserved.
+ *
+ * This is part of the Xively C Client library,
+ * it is licensed under the BSD 3-Clause license.
+ */
+
+#include <xi_bsp_io_net.h>
+
+//#include <sys/select.h>
+#include <lwip/netdb.h>
+
+#include <string.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef MAX
+#define MAX( a, b ) ( ( ( a ) > ( b ) ) ? ( a ) : ( b ) )
+#endif
+
+xi_bsp_io_net_state_t xi_bsp_io_net_create_socket( xi_bsp_socket_t* xi_socket )
+{
+    *xi_socket = socket( AF_INET, SOCK_STREAM, 0 );
+
+    if ( -1 == *xi_socket )
+    {
+        return XI_BSP_IO_NET_STATE_ERROR;
+    }
+
+    /* Enable nonblocking mode for a posix socket */
+    const int flags = fcntl( *xi_socket, F_GETFL, 0 );
+
+    if ( -1 == flags || -1 == fcntl( *xi_socket, F_SETFL, flags | O_NONBLOCK ) )
+    {
+        return XI_BSP_IO_NET_STATE_ERROR;
+    }
+
+    return XI_BSP_IO_NET_STATE_OK;
+}
+
+xi_bsp_io_net_state_t
+xi_bsp_io_net_connect( xi_bsp_socket_t* xi_socket, const char* host, uint16_t port )
+{
+    struct hostent* hostinfo = gethostbyname( host );
+    int errno_cpy = 0; /* errno is reset every time it's read */
+
+    /* if null it means that the address has not been found */
+    if ( NULL == hostinfo )
+    {
+        return XI_BSP_IO_NET_STATE_ERROR;
+    }
+
+    struct sockaddr_in name = {.sin_family = AF_INET,
+                               .sin_port   = htons( port ),
+                               .sin_addr =
+                                   *( ( struct in_addr* )hostinfo->h_addr_list[0] ),
+                               .sin_zero = {0}};
+
+    if ( -1 ==
+         connect( *xi_socket, ( struct sockaddr* )&name, sizeof( struct sockaddr ) ) )
+    {
+        errno_cpy = errno;
+        return ( EINPROGRESS == errno_cpy ) ? XI_BSP_IO_NET_STATE_OK
+                                            : XI_BSP_IO_NET_STATE_ERROR;
+    }
+    else
+    {
+        // todo_atigyi: what to do here?
+        // does this mean the socket is BLOCKING?
+        return XI_BSP_IO_NET_STATE_ERROR;
+    }
+
+    return XI_BSP_IO_NET_STATE_ERROR;
+}
+
+xi_bsp_io_net_state_t xi_bsp_io_net_connection_check( xi_bsp_socket_t xi_socket,
+                                                      const char* host,
+                                                      uint16_t port )
+{
+    ( void )host;
+    ( void )port;
+
+    int valopt    = 0;
+    socklen_t lon = sizeof( int );
+
+    // printf( "getsockopt\n" );
+
+    int result =
+        getsockopt( xi_socket, SOL_SOCKET, SO_ERROR, ( void* )( &valopt ), &lon );
+
+    if ( result < 0 )
+    {
+        // int errval = errno;
+        // printf( "getsockopt failed with ret %d, errval %d", result, errval );
```

Comment:
remove?

Reply:
I'm leaving this one in, if you don't mind. Seems quite useful for BSP debugging.